pdata/ptrace/jsoniter_test.go

```diff
@@ -0,0 +1,401 @@
+	link.Attributes().UpsertBool("bool", true)
+	link.Attributes().UpsertInt("int", 1)
+	link.Attributes().UpsertDouble("double", 1.1)
+	link.Attributes().UpsertBytes("bytes", []byte("foo"))
+	// Add another span.
+	sp2 := il.Spans().AppendEmpty()
+	sp2.SetName("testSpan2")
+	return td
+}()
+
+func TestJSONIter(t *testing.T) {
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalTraces(tracesOTLPFull)
+	assert.NoError(t, err)
+
+	decoder := newJSONIterUnmarshaler()
+	got, err := decoder.UnmarshalTraces(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, tracesOTLPFull, got)
+}
+
+func BenchmarkJSONUnmarshal(b *testing.B) {
+	b.ReportAllocs()
+
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalTraces(tracesOTLPFull)
+	assert.NoError(b, err)
+	decoder := newJSONUnmarshaler()
+
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_, err := decoder.UnmarshalTraces(jsonBuf)
+			assert.NoError(b, err)
+		}
+	})
+}
+
+func BenchmarkTracesJSONiterUnmarshal(b *testing.B) {
+	b.ReportAllocs()
+
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalTraces(tracesOTLPFull)
+	assert.NoError(b, err)
+	decoder := newJSONIterUnmarshaler()
+
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_, err := decoder.UnmarshalTraces(jsonBuf)
+			assert.NoError(b, err)
+		}
+	})
+}
+
+func TestReadInt64(t *testing.T) {
+	var data = `{"intAsNumber":1,"intAsString":"1"}`
+	iter := jsoniter.ConfigFastest.BorrowIterator([]byte(data))
+	iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+		switch f {
+		case "intAsNumber":
+			v := readInt64(iter)
+			assert.Equal(t, int64(1), v)
+		case "intAsString":
+			v := readInt64(iter)
+			assert.Equal(t, int64(1), v)
+		}
+		return true
+	})
+	assert.NoError(t, iter.Error)
+}
+
+func Test_readTraceData(t *testing.T) {
+	t.Run("unknown field", func(t *testing.T) {
+		jsonStr := `{"extra":""}`
+		iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
+		defer jsoniter.ConfigFastest.ReturnIterator(iter)
+		readTraceData(iter)
+		if assert.Error(t, iter.Error) {
+			assert.Contains(t, iter.Error.Error(), "unknown field")
+		}
+	})
+}
+
+func Test_readResourceSpans(t *testing.T) {
+	t.Run("unknown field", func(t *testing.T) {
+		jsonStr := `{"extra":""}`
+		iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
+		defer jsoniter.ConfigFastest.ReturnIterator(iter)
+		readResourceSpans(iter)
+		if assert.Error(t, iter.Error) {
+			assert.Contains(t, iter.Error.Error(), "unknown field")
+		}
+	})
+	t.Run("unknown resource field", func(t *testing.T) {
+		jsonStr := `{"resource":{"extra":""}}`
+		iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
+		defer jsoniter.ConfigFastest.ReturnIterator(iter)
+		readResourceSpans(iter)
+		if assert.Error(t, iter.Error) {
+			assert.Contains(t, iter.Error.Error(), "unknown field")
+		}
+	})
+}
+
+func Test_readInstrumentationLibrarySpans(t *testing.T) {
```

Comment:
Nice tests overall, but we apply a different "design". For this test we would have 2 tests:

```golang
func TestReadInstrumentationLibrarySpansUnkownField(t *testing.T) {
	jsonStr := `{"extra":""}`
	iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
	defer jsoniter.ConfigFastest.ReturnIterator(iter)
	readInstrumentationLibrarySpans(iter)
	if assert.Error(t, iter.Error) {
		assert.Contains(t, iter.Error.Error(), "unknown field")
	}
}

func TestReadInstrumentationLibrarySpansUnkownInstrumentationLibraryField(t *testing.T) {
	jsonStr := `{"instrumentationLibrary":{"extra":""}}`
	iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
	defer jsoniter.ConfigFastest.ReturnIterator(iter)
	readInstrumentationLibrarySpans(iter)
	if assert.Error(t, iter.Error) {
		assert.Contains(t, iter.Error.Error(), "unknown field")
	}
}
```

Reply:
OK, I need some time to update all.

Reply:
Done